solace/src/main/java/org/apache/apex/malhar/solace/AbstractSolaceBaseInputOperator.java

```diff
@@ -0,0 +1,319 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.apex.malhar.solace;
+
+import java.io.IOException;
+import java.util.LinkedList;
+import java.util.concurrent.ArrayBlockingQueue;
+import java.util.concurrent.BlockingQueue;
+import javax.validation.constraints.NotNull;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.solacesystems.jcsmp.BytesXMLMessage;
+import com.solacesystems.jcsmp.Consumer;
+import com.solacesystems.jcsmp.JCSMPChannelProperties;
+import com.solacesystems.jcsmp.JCSMPException;
+import com.solacesystems.jcsmp.JCSMPFactory;
+import com.solacesystems.jcsmp.JCSMPProperties;
+import com.solacesystems.jcsmp.JCSMPReconnectEventHandler;
+import com.solacesystems.jcsmp.JCSMPSession;
+import com.solacesystems.jcsmp.SessionEvent;
+import com.solacesystems.jcsmp.SessionEventArgs;
+import com.solacesystems.jcsmp.SessionEventHandler;
+import com.solacesystems.jcsmp.XMLMessageListener;
+import com.datatorrent.api.Context;
+import com.datatorrent.api.InputOperator;
+import com.datatorrent.api.Operator;
+import com.datatorrent.api.Operator.CheckpointListener;
+import com.datatorrent.common.util.BaseOperator;
+import com.datatorrent.lib.io.IdempotentStorageManager;
+import com.datatorrent.netlet.util.DTThrowable;
+
+@SuppressWarnings("unused")
+public abstract class AbstractSolaceBaseInputOperator<T> extends BaseOperator implements
+    InputOperator, Operator.ActivationListener<Context.OperatorContext>, CheckpointListener
+{
+
+  private static final Logger logger = LoggerFactory.getLogger(AbstractSolaceBaseInputOperator.class);
+
+  @NotNull
+  protected JCSMPProperties properties = new JCSMPProperties();
```

Comment:
We should take a look at this properties variable and its NotNull constraint and where all this variable is set. My concerns are:

- here it is initialized with an empty JCSMProperties object only to address the NotNull constraint
- the setProperties() method just overwrites the existing variable with a new value
- there is a place where a property is set in the existing JCSMProperties object JCSMPProperties.REAPPLY_SUBSCRIPTIONS.
- if setProperties() is called after setReapplySubscriptions() won't it clobber the earlier setting?

To solve this can we change setProperties() to copy properties from the argument to the member variable as my comment there says?

Reply:
Having the properties variable initialized helps in configuring the operator from properties from a property file, for example <name>operator.prop.properties.properties['name']</name><value>...</value>